File: TypeScript/app/gilded-rose.ts
import { ICustomItem } from './interfaces';
import {
  AgedBrie,
  BackstagePasses,
  ConjuredItem,
  Item,
  LegendaryItem,
  RegularItem,
} from './inventory';

export class GildedRose {
  
  // stores the respective custom class for each esoteric item
  // removes hard coded logic from getCustomItem()
  private static itemClassReference = [
    {
      name: 'Backstage passes to a TAFKAL80ETC concert',
      class: BackstagePasses,
    },
    { name: 'Aged Brie', class: AgedBrie },
    { name: 'Sulfuras, Hand of Ragnaros', class: LegendaryItem },
    { pattern: 'Conjured', class: ConjuredItem },
  ];
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  // find matching class based on full name match or partial pattern match ('Conjured')
  // otherwise return RegularItem instance
  getCustomItem(item: Item): ICustomItem {
    const itemClass =
      GildedRose.itemClassReference.find(
        ic =>
          item.name === ic.name ||
          (ic.pattern && item.name.includes(ic.pattern)),
      )?.class || RegularItem;
    return new itemClass(item);
  }

  updateQuality() {
    return this.items.map(item => {
      const customItem = this.getCustomItem(item);
      customItem.update()
      return customItem.item;
    });
  }
}

File: ruby/gilded_rose.rb
require_relative 'inventory'



class GildedRose
  # stores the respective custom class for each esoteric item
  # removes hardcoded logic from getCustomItem()
  @item_class_reference = [
    {
      name: 'Backstage passes to a TAFKAL80ETC concert',
      ref: BackstagePasses,
    },
    { name: 'Aged Brie', ref: AgedBrie },
    { name: 'Sulfuras, Hand of Ragnaros', ref: LegendaryItem },
    { pattern: 'Conjured', ref: ConjuredItem },
  ]

  class << self
    attr_reader :item_class_reference
  end

  def initialize(items)
    @items = items
  end

  #find matching class based on full name patch or partial pattern match ('Conjured')
  # otherwise return RegularItem instance
  def get_custom_item(item)
    class_ref = GildedRose.item_class_reference.find do |ic|
      item.name == ic[:name] || (ic[:pattern] && item.name.include?(ic[:pattern]))
    end
    custom_class = class_ref.nil? ? RegularItem : class_ref[:ref]
    custom_class.new(item)
  end

  def update_quality()
    @items.each do |item|
      custom_item = get_custom_item(item)
      quality_change = custom_item.rate_of_quality_change
      custom_item.update(quality_change)
    end
  end
end

class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end
